x-pack/plugins/security_solution/server/lib/detection_engine/signals/search_after_bulk_create.ts

```diff
@@ -47,183 +51,203 @@ export const searchAfterAndBulkCreate = async ({
-          timestampOverride: ruleParams.timestampOverride,
+    while (totalToFromTuples.length > 0) {
+      const tuple = totalToFromTuples.pop();
+      if (tuple == null || tuple.to == null || tuple.from == null) {
+        logger.error(buildRuleMessage(`[-] malformed date tuple`));
+        return createSearchAfterReturnType({
+          success: false,
+          errors: ['malformed date tuple'],
         });
-        toReturn = mergeReturns([
-          toReturn,
-          createSearchAfterReturnTypeFromResponse({
-            searchResult,
-            timestampOverride: ruleParams.timestampOverride,
-          }),
-          createSearchAfterReturnType({
-            searchAfterTimes: [searchDuration],
-            errors: searchErrors,
-          }),
-        ]);
-        // determine if there are any candidate signals to be processed
-        const totalHits = createTotalHitsFromSearchResult({ searchResult });
-        logger.debug(buildRuleMessage(`totalHits: ${totalHits}`));
-        logger.debug(
-          buildRuleMessage(`searchResult.hit.hits.length: ${searchResult.hits.hits.length}`)
-        );
-
-        // search results yielded zero hits so exit
-        // with search_after, these two values can be different when
-        // searching with the last sortId of a consecutive search_after
-        // yields zero hits, but there were hits using the previous
-        // sortIds.
-        // e.g. totalHits was 156, index 50 of 100 results, do another search-after
-        // this time with a new sortId, index 22 of the remaining 56, get another sortId
-        // search with that sortId, total is still 156 but the hits.hits array is empty.
-        if (totalHits === 0 || searchResult.hits.hits.length === 0) {
-          logger.debug(
-            buildRuleMessage(
-              `${
-                totalHits === 0 ? 'totalHits' : 'searchResult.hits.hits.length'
-              } was 0, exiting and moving on to next tuple`
-            )
-          );
-          break;
-        }
+      }
+      signalsCreatedCount = 0;
+      while (signalsCreatedCount < tuple.maxSignals) {
+        try {
+          logger.debug(buildRuleMessage(`sortIds: ${sortId}`));
 
-        // filter out the search results that match with the values found in the list.
-        // the resulting set are signals to be indexed, given they are not duplicates
-        // of signals already present in the signals index.
-        const filteredEvents = await filterEventsAgainstList({
-          listClient,
-          exceptionsList,
-          logger,
-          eventSearchResult: searchResult,
-          buildRuleMessage,
-        });
+          // timestampsAndIndices can be empty if the rule author doesn't have
+          // view_index_metadata privileges and we can't access field mappings api
+          const indexPatternsToSearch = !isEmpty(timestampsAndIndices)
```

Comment:
In testing I found that if a rule queries across an index that has the timestamp override field but not `@timestamp` and an index that has neither `@timestamp` nor the override field (e.g. `myfakeindex-1` and `myfakeindex-2` from the PR description) then `indexPatternsToSearch` becomes an empty array when timestamp=`@timestamp` for this loop. The empty array causes the query to search across all ES indices.

Reply:
goooooood catch